tokio-util/src/task/join_map.rs

```diff
@@ -0,0 +1,499 @@
+    /// will be cancelled and replaced with the new one.
+    ///
+    /// # Panics
+    ///
+    /// This method panics if it is called outside of a `LocalSet`.
+    ///
+    /// [`LocalSet`]: tokio::task::LocalSet
+    pub fn spawn_local<F>(&mut self, key: K, task: F)
+    where
+        F: Future<Output = V>,
+        F: 'static,
+    {
+        let task = self.joins.spawn_local(mk_task(&key, task));
+        self.insert(key, task);
+    }
+
+    /// Spawn the provided task on the provided [`LocalSet`] and store it in
+    /// this `JoinMap` with the provided key.
+    ///
+    /// If a task previously existed in the `JoinMap` for this key, that task
+    /// will be cancelled and replaced with the new one.
+    ///
+    /// [`LocalSet`]: tokio::task::LocalSet
+    pub fn spawn_local_on<F>(&mut self, key: K, task: F, local_set: &LocalSet)
+    where
+        F: Future<Output = V>,
+        F: 'static,
+    {
+        let task = self.joins.spawn_local_on(mk_task(&key, task), local_set);
+        self.insert(key, task)
+    }
+
+    fn insert(&mut self, key: K, abort: AbortHandle) {
+        if let Some(prev) = self.aborts.insert(key, abort) {
+            prev.abort();
+        }
+    }
+
+    /// Waits until one of the tasks in the map completes and returns its
+    /// output, along with the key corresponding to that task.
+    ///
+    /// Returns `None` if the map is empty.
+    ///
+    /// # Cancel Safety
+    ///
+    /// This method is cancel safe. If `join_one` is used as the event in a `tokio::select!`
+    /// statement and some other branch completes first, it is guaranteed that no tasks were
+    /// removed from this `JoinMap`.
+    pub async fn join_one(&mut self) -> Result<Option<(K, V)>, JoinError> {
+        match self.joins.join_one().await {
+            Ok(Some((key, val))) => {
+                self.aborts.remove(&key);
+                Ok(Some((key, val)))
+            }
+            Ok(None) => Ok(None),
+            // If a task panics or is aborted, we must clear its `AbortHandle`
+            // out of the map of abort handles, as the `AbortHandle` keeps the
+            // task from being deallocated.
+            Err(e) => {
+                // XXX(eliza): i don't _love_ the `retain` here; it would be
+                // nice if we could just look up the individiual task and remove
+                // *it* from the map. but, we don't have the key in this case.
+                // perhaps we could instead add the ability to compare
+                // `AbortHandle`s and `JoinHandle`s to see if they correspond to
+                // the same task, and change to some kind of map type allowing
+                // inverse lookups...but that would also mean adding `Hash` and `Eq`
+                // commitments to the `JoinHandle`/`AbortHandle` APIs, which i'm not
+                // sure if we want to do...
+                self.aborts.retain(|_, task| task.is_active());
+                Err(e)
```

Comment:
I don't like this either.

Reply:
c4f6b7ff263d5681f82241f69f0360de74eb1602 gets rid of this.